### PR TITLE
Donation factory 3rd attempt

### PR DESCRIPTION
**Pull request by helgatheviking**
Some days I really suck at git. Follow up to #98

### CODE REVIEW

File: includes/donations/class-charitable-donation-factory.php [hunk 1]
```diff
@@ -0,0 +1,79 @@
...
+ * @category	Class
+ * @author 		Eric Daams
+ */
+class Charitable_Donation_Factory {
+
+	/**
+	 * Get donation.
+	 *
+	 * @param bool $the_donation (default: false)
+	 * @return WC_Order|bool
```
Comment (ericnicolaas) [inline includes/donations/class-charitable-donation-factory.php]: We're not in WooCommerce :)

File: includes/donations/class-charitable-donation-factory.php [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+	 */
+	public function get_donation( $the_donation = false ) {
+		global $post;
+
+		if ( false === $the_donation ) {
+			$the_donation = $post;
+		} elseif ( is_numeric( $the_donation ) ) {
+			$the_donation = get_post( $the_donation );
+		} elseif ( $the_donation instanceof WC_Order ) {
```
Comment (ericnicolaas) [inline includes/donations/class-charitable-donation-factory.php]: We're not in WooCommerce :)

File: includes/donations/class-charitable-donation-factory.php (hunk 1 above)
Comment (helgatheviking) [inline includes/donations/class-charitable-donation-factory.php]: true. I probably need a good grunt replace to make sure to catch all WooCommerce strings.

File: includes/donations/class-charitable-donation-factory.php (hunk 2 above)
Comment (helgatheviking) [inline includes/donations/class-charitable-donation-factory.php]: just pushed a commit that should change these 2 line notes.

**Comment by ericnicolaas** [comment]
With the logging functions & methods, either the class methods should call the functions, or the functions should call the class methods. I think it probably makes most sense for the class methods to call the functions:

```
public function get_donation_log( $donation_id = null ) {
    if(  $donation_id ){
        _deprecated_argument( __METHOD__, '1.3.0', '$donation_id no longer required as get_donation_log() is used in object context. Use $donation->get_donation_log()' );
    }

    return charitable_get_donation_log( $this->donation_id );
}
```

**Comment by helgatheviking** [comment]
But why bother with the functions? I thought they were there originally, but they aren't so I just added duplication.

**Comment by ericnicolaas** [comment]
Alternatively, we could get rid of the functions, but the one implication of that is that you then have to create a Donation object if all you want is a log or to update a log.

**Comment by helgatheviking** [comment]
Does creating a new object have a lot of overhead? It is getting the post via `get_post` in the constructor. Is it being used in places where the donation object would not already exist? Otherwise, if we have to keep them then I agree that it makes sense to have the methods call the functions.

**Comment by ericnicolaas** [comment]
OK, I have started work on merging this in. Already made some changes, which are pushed up to https://github.com/Charitable/Charitable/tree/pr/108

**Comment by helgatheviking** [comment]
Great! I've been finding little tweaks the more I use it with Recurring.

**Comment by ericnicolaas** [comment]
@helgatheviking OK, made a few more tweaks, but merged into `master` now, with a view to releasing this for 1.4.0.

**Comment by helgatheviking** [comment]
Awesome! I'm flying to France tomorrow night, but once I get settled should
be back at it. Moving the subscription creation to after the donation
creation will go a long way to getting that closer to finished.

**Pull request closed without merging**